### ISSUE TITLE
File Permissions Don't Survive Local Source Build

### ISSUE THREAD

**Issue by berwyn** [Azure/azure-acr-plugin#40]
I'm building my company's Rails app, and uploading the code from the Jenkins server to ACR. The build succeeds and everything works well... until we run the image. At some point, the execute bit is stripped off of `bin/*` in our repo, leaving our scripts unable to run. Based on a loose guess, it seems to be coming from compressing the repo into the tarball before uploading.

**Comment by gavinfish** [comment]
@yuwzho any idea about the compressing and uploading details here?

**Comment by yuwzho** [comment]
@berwyn Do you have more detail information? Like giving a small sample, or package the tarball file yourself and give it to ACR.

**Comment by berwyn** [comment]
Yeah, I ended up exposing a PAT to the build runtime and using the Git option and that correctly keeps the executable bit. As for a small sample, the output of `rails new` with default arguments would provide a reasonable equivalent. To be specific, you'd end up with a tree like this:

```
- (app)
|- bin/
| |- rails
| |- bundle
| |- rake
| `- (others)
|- app/
|- spec/
( and more)
```

Of note, everything in `bin/` has the executable bit set, (`-rwxr-xr-x` or something similar).

We have, for our app, also included a `scripts` folder with bash scripts

```
- (app)
|- bin/
|- app/
|- spec/
|- scripts/
| |- docker-rubocop.sh
| `- (others)
(etc)
```

These scripts all also have the executable bit (`-rwxr-xr-x`) set.

If I queue the build via the plugin as such:
```groovy
acrQuickTask(
  azureCredentialsId: 'jenkins-az-sp',
  resourceGroupName: 'myResourceGroup',
  registryName: 'myRegistry',
  platform: 'Linux',
  dockerfile: 'Dockerfile,
  imageNames: [[image: 'myregistry.azurecr.io/myrepo:branch']]
)
```

Then the repo will get compressed and the Dockerfile builds just fine. However, once the Dockerfile is run, in this case trying to execute one of the aforementioned scripts, I get permission denied errors because the executable bit has been stripped, leaving the scripts as only `rw` (`-rw-r--r--`):
```
/home/jenkins/workspace/<repo>@tmp/durable-47253cdb/script.sh: 8: /home/jenkins/workspace/<repo>@tmp/durable-47253cdb/script.sh: scripts/docker-rubocop.sh: Permission denied
```

This also applies to the binstubs that Rails uses (located in `/bin`), which leads to the server being unable to boot as well.

To be honest, I'm not sure if this is a bug or intentional, but if it's intentional that no attempt is made to preserve the unix permissions then this should probably be documented so that people like myself know to either create the tarball themselves or to use the git methodology (which does preserve the unix permissions)

**Comment by yuwzho** [comment]
The compress logic doesn't modify any permission at all. I may need some time to dig it out. @berwyn

**Comment by berwyn** [comment]
No need to rush on my account. I was able to work around it using the git builder and that's working excellently!
